node/network/collator-protocol/src/collator_side.rs

```diff
@@ -91,6 +90,96 @@ impl metrics::Metrics for Metrics {
+	/// All [`ValidatorId`]'s that are assigned to us in this group.
+	validator_ids: HashSet<ValidatorId>,
+	/// The mapping from [`PeerId`] to [`ValidatorId`]. This is filled over time as we learn the [`PeerId`]'s from the
+	/// authority discovery. It is not ensured that this will contain *all* validators of this group.
+	peer_ids: HashMap<PeerId, ValidatorId>,
+	/// All [`ValidatorId`]'s of the current group to that we advertised our collation.
+	advertised_to: HashSet<ValidatorId>,
+}
+
+impl ValidatorGroup {
+	/// Returns `true` if we should advertise our collation to the given peer.
+	fn should_advertise_to(&self, peer: &PeerId) -> bool {
+		match self.peer_ids.get(peer) {
+			Some(validator_id) => !self.advertised_to.contains(validator_id),
+			None => false,
+		}
+	}
+
+	/// Should be called after we advertised our collation to the given `peer` to keep track of it.
+	fn advertised_to_peer(&mut self, peer: &PeerId) {
+		if let Some(validator_id) = self.peer_ids.get(peer) {
+			self.advertised_to.insert(validator_id.clone());
+		}
+	}
+
+	/// Add a [`PeerId`] that belongs to the given [`ValidatorId`].
+	///
+	/// This returns `true` if the given validator belongs to this group and we could insert its [`PeerId`].
+	fn add_peer_id_for_validator(&mut self, peer_id: &PeerId, validator_id: &ValidatorId) -> bool {
+		if !self.validator_ids.contains(validator_id) {
+			false
+		} else {
+			self.peer_ids.insert(peer_id.clone(), validator_id.clone());
+			true
+		}
+	}
+}
+
+impl From<HashSet<ValidatorId>> for ValidatorGroup {
+	fn from(validator_ids: HashSet<ValidatorId>) -> Self {
+		Self {
+			validator_ids,
+			peer_ids: HashMap::new(),
+			advertised_to: HashSet::new(),
+		}
+	}
+}
+
+/// Wrapper around the connection requests per relay parent.
+///
+/// Implements [`Future`] to check if any of the internal requests returned a result to propagate it to the calling
+/// context. It is safe to call this future multiple times, if there are no requests or none of them returned a result
+/// (because they returned either [`Poll::Pending`]/`None`) it just returns [`Poll::Pending`].
+#[derive(Default)]
+struct ConnectionRequests {
+	requests: HashMap<Hash, validator_discovery::ConnectionRequest>,
+}
+
+impl ConnectionRequests {
+	/// Insert a new connection request.
+	fn put(&mut self, relay_parent: Hash, request: validator_discovery::ConnectionRequest) {
+		self.requests.insert(relay_parent, request);
+	}
+
+	/// Remove a connection request for the given `relay_parent`.
+	fn remove(&mut self, relay_parent: &Hash) {
+		self.requests.remove(&relay_parent);
+	}
+}
+
+impl Future for ConnectionRequests {
+	type Output = (Hash, ValidatorId, PeerId);
+
+	fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+		for (relay_parent, request) in self.requests.iter_mut() {
+			if let Poll::Ready(Some(res)) = request.poll_next_unpin(cx) {
+				return Poll::Ready((*relay_parent, res.0, res.1))
+			}
+		}
+
+		Poll::Pending
```

Comment:
That breaks the [`poll` contract](https://doc.rust-lang.org/std/future/trait.Future.html#tymethod.poll)

This function returns:

    Poll::Pending if the future is not ready yet
    Poll::Ready(val) with the result val of this future if it finished successfully.

Once a future has finished, clients should not poll it again.

Reply:
This should be an infinite stream rather than a future.

Reply:
I know what this does and I thought about documenting it. However, I removed the docs again. If this would be public, I would have made this more clear. But here we don't rely on being waken up by this future.

Reply:
I find it more weird when I see such an explicit call.